IGNITE-12291. Add MergeSort distributed cache query reducer.

timoninmaxim opened the pull request:
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java:
```diff
@@ -169,103 +97,30 @@ protected GridCacheDistributedQueryFuture(GridCacheContext<K, V> ctx, long reqId
 
     /** {@inheritDoc} */
     @Override public boolean onCancelled() {
-        firstPageLatch.countDown();
+        reducer.onLastPage();
 
         return super.onCancelled();
     }
 
     /** {@inheritDoc} */
     @Override public void onTimeout() {
-        firstPageLatch.countDown();
+        reducer.onLastPage();
 
         super.onTimeout();
     }
```
Do we need all these callbacks in QueryFutureAdapter just to delegate calls to reducer
and having reducer() at same time?

timoninmaxim replied:
@AMashenkov hi! Thanks for have a look this PR. Agree, I'll move it to the base class (`GridCacheQueryFutureAdapter`), there are additional functionality on this callbacks methods (onCancelled, onTimeout)

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryResultFetcher.java:
```diff
@@ -0,0 +1,243 @@
+    private final IgniteLogger log;
+
+    /** */
+    CacheQueryResultFetcher(
+        final GridCacheContext cctx,
+        final ConcurrentMap<Long, GridCacheDistributedQueryFuture<?, ?, ?>> qryFuts) {
```
CacheQueryResultFetcher doesn't look like good class name.
1. Actually, it is not fetcher for particular query, it is a holder for object-fetchers for multiple queries.
2. Why it depends on qryFuts?
3. The functionality of the class is limited to create-and-send requests, but any response processing is out of scope...
So, it is a 'requester', but not 'fetcher'

timoninmaxim replied:
Agree, that the "fetcher" name isn't good. I think it's much as Trigger or Requester. 

Moment to load page is depends on reducer implementation, so reducer is responsible for triggering pages. Reducer is created per request, and fetcher is single per Cache, so we can hold a reference with the same lifecycle.  

Alternative is to store a future in reducer (and we can do it, as get fut in constructor), and provide it to fetcher instead of reqId. Also it help us to get rid of query qryFuts. Looks OK for me. I'll do this.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheEntryWithPayload.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        return "CacheEntryWithPayload{" +
+            "entry=" + super.toString() +
+            ", payload=" + payload +
+            '}';
+    }
```
```suggestion
    @Override public String toString() {
        return S.toString(CacheEntryWithPayload.class, this);
    }
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,232 @@
+            sendLocal(req);
+    }
+
+    /** Send and handle request to local node. */
```
```suggestion
    /** 
    * Send and handle request to local node. 
    */
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/AbstractCacheQueryReducer.java:
```diff
@@ -0,0 +1,183 @@
+    private final long timeoutTime;
+
+    /** Lock shared between this reducer and future. */
+    private final Object sharedLock;
```
```suggestion
    private final Object queueLock;
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/AbstractCacheQueryReducer.java:
```diff
@@ -0,0 +1,183 @@
+        sharedLock = fut.lock;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Object sharedLock() {
+        return sharedLock;
+    }
```
This looks like abstraction leakage antipattern.

timoninmaxim replied:
@AMashenkov hi! Thanks for review. I've removed this method from interface, and make protected.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheEntryWithPayload.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /** */
+    public Object payload() {
```
```suggestion
    public P payload() {
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,236 @@
+
+        assert qryMgr != null;
+
+        try {
+            final GridCacheQueryRequest req = new GridCacheQueryRequest(cctx.cacheId(),
+                reqId,
+                fieldsQry,
+                cctx.startTopologyVersion(),
+                cctx.deploymentEnabled());
+
+            // Process cancel query directly (without sending) for local node.
+            sendLocal(req);
+
+            if (!nodes.isEmpty()) {
+                for (ClusterNode node : nodes) {
+                    try {
+                        cctx.io().send(node, req, cctx.ioPolicy());
+                    }
+                    catch (IgniteCheckedException e) {
+                        if (cctx.io().checkNodeLeft(node.id(), e, false)) {
+                            if (log.isDebugEnabled())
+                                log.debug("Failed to send cancel request, node failed: " + node);
+                        }
+                        else
+                            U.error(log, "Failed to send cancel request [node=" + node + ']', e);
+                    }
+                }
+            }
+        }
```
```suggestion
        if (nodes.isEmpty())
            return;
            
        try {
            final GridCacheQueryRequest req = new GridCacheQueryRequest(cctx.cacheId(),
                reqId,
                fieldsQry,
                cctx.startTopologyVersion(),
                cctx.deploymentEnabled());

            // Process cancel query directly (without sending) for local node.
            sendLocal(req);

            for (ClusterNode node : nodes) {
                try {
                    cctx.io().send(node, req, cctx.ioPolicy());
                }
                catch (IgniteCheckedException e) {
                    if (cctx.io().checkNodeLeft(node.id(), e, false)) {
                        if (log.isDebugEnabled())
                            log.debug("Failed to send cancel request, node failed: " + node);
                     }
                    else
                        U.error(log, "Failed to send cancel request [node=" + node + ']', e);
                }
            }
        }
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/UnsortedDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
```
Should this method countdown firstPageLatch?

timoninmaxim replied:
No, as `cancel()` always prefaced with `onLastPage()` call. 

Also, I checked that actually the `awaitFirstItem()` used only for ScanQuery only in case user specified partitions to enable retry this query if first page loading failed. Looks too narrow case, and I think there should be a separate activity to investigate this functionality.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,298 @@
+    ) {
+        super(fut, reqId, fetcher, nodes);
+
+        synchronized (queueLock()) {
```
Why we need any synchronization here?
No statements have any side effects.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,298 @@
+
+    /** {@inheritDoc} */
+    @Override public void addPage(@Nullable UUID nodeId, Collection<R> data) {
+        assert Thread.holdsLock(queueLock());
```
Why do we care about external synchronization?
Seems,  it is enough to avoid NPE after streamsMap.get(nodeId).

timoninmaxim replied:
You're right. We need to check streamsMap.get() for NPE in case of concurrent `onProcessQueryResponse()` and `onNodeLeft()`. Guarding with lock doesn't help us in this case.

timoninmaxim replied:
You're right. I've simplified code, as nodeId == null for distributed reducer means only error. So we can remove check for sync (that was required to guard streamsMap) and replace with check for NPE.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,298 @@
+            if (!streamsMap.containsKey(nodeId)) {
+                assert data.isEmpty();
+
+                for (PageStream stream: streamsMap.values())
+                    stream.addPage(data);
```
Why do you pass 'data' to all streams, if got a page from an unknown node?

timoninmaxim replied:
In this case we know that nodeId set to NULL only in case of error, that data is empty. So it's safe to put data. Note that we put this to all nodes to faster finish stream traverse to faster check future for error.

AMashenkov replied:
So, why we have a magic constant NULL instead of a separate callback for this with error as a param?
It is totally unclear, why UnsortedReducer doesn't have a similar null check...

Also, passing an empty page looks like a magic constant as well.
Do we need to notify other streams? Do we know at this point, that prefetched result (failed) will ever be needed?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/UnsortedDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,211 @@
+    /** {@inheritDoc} */
+    @Override public boolean onNodeLeft(UUID nodeId) {
+        synchronized (queueLock()) {
+            return subgrid.contains(nodeId);
```
Here we know the node was left.
Is there any reason to return smth to higher-level instead of releasing a latch instantly?
It is non-obvious if onPage will ever be called.

timoninmaxim replied:
You're correct, it fut.onPage(error) can be invoked in this method. But it looks like as spaghetti. Stack of calls is:
fut.onNodeLeft -> rdc.onNodeLeft -> fut.onPage(error) -> rdc.onPage. 

It's not too much, but looks strange. 

> It is non-obvious if onPage will ever be called.

Why do you think? JavaDocs clearly covers return value of rdc.onNodeLeft.

AMashenkov replied:
You can have 2 separate calls instead.
fut.onNodeLeft -> rdc.onNodeLeft 
fut.onNodeLeft -> rdc.isFailed/isFinished

timoninmaxim replied:
Renamed method, now it just checks that specified node is running query.

> Is there any reason to return smth to higher-level instead of releasing a latch instantly?

I tried to avoid this, but found that we must release a latch (I mean CountDownLatch) after we mark future as done. It's required for correct return exception to a user in case it's hanging on `awaitInitialization()` method.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,298 @@
+
+            // Skip if stream has head: on previous next() head extracts from stream[streamOff], others streams do
+            // have head or it is the first run.
+            if (s.head() != null)
+                break;
```
AFAIK, at the very first call, first pages will be requested from all data nodes.
The next data will be got from the head.
When the first page data will be consumed, then the next page will be requested synchronously.

Thus, prefetch only works for the first nodes pages only.

timoninmaxim replied:
Next pages are requested asynchronously in moment after preparing a head for a stream and finding that queue is empty. We compare heads and bubble them on every user's `next()`. So for multiple streams (1 stream per node), for even-distributed between nodes data we will get head from new stream for every next() call. Then we can expect that page loading is non-blocking in most cases.

timoninmaxim replied:
Check this code:
https://github.com/apache/ignite/pull/9081/files#diff-068da2d055341b4686f2dc28d687b5c58678a255cde712e4062588bf0fc4f46fR135

AMashenkov replied:
After you got the last item from the first page, you will be blocked on `next()` call until the second page will be received from the node, otherwise you will not be able to resort streams correctly.
You need to have at least one page from every node to be ready at any time for non-blocking reducing.

I wouldn't bet on equal data distribution. Assume, one node will have the largest part of the query result,
then you will end up in synchronous ping-pong with a single node.

Effective prefetching is to request next page right after you start processing the previous one.
If you bother about fetching few additional pages, then you can configure twice smaller page, aren't you?

timoninmaxim replied:
> Effective prefetching is to request next page right after you start processing the previous one

It is how it's done now. By "queue is empty" I mean the queue variable in PageStream, that stores page's collections. When the queue is empty it means that we start processing last page (and data moved from queue to iterator [1]) and we request a new page in parallel. 

By the link above I trigger this load right after moving page to iterator, check the condition: queue.size() < 1. So queue is empty but we still have data in iterator.

We can change this condition to queue.size() < 2, so we will have at least 2 pages in any moment (1 in iterator, 1 in a queue). So it can be useful if user processing page in iterator faster then remote node responses.

Am I right that you propose to do this (increment 1 to 2)? Should we do it configurable?

[1] https://github.com/apache/ignite/pull/9081/files#diff-068da2d055341b4686f2dc28d687b5c58678a255cde712e4062588bf0fc4f46fR75

timoninmaxim replied:
Reworked this, now there is a single logic for loading pages for PageStream and NodePageStream. After moving last item from queue to iterator, we trigger load page.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,290 @@
+
+        List<UUID> nodes;
+
+        synchronized (queueLock()) {
```
You use the same lock to synchronize access to streams[] and all queues in all NodeStrems objects.
What is the purpose to use a shared lock?

timoninmaxim replied:
In master branch (single stream) one of purposes of the lock is to guard consistency of `queue` and collections `rcvd`, `subgrid`. As far as I undertstand it's required to avoid excess requests to nodes that already finish query. It could lead to data duplication in final result, as query runs from beginning . Actually as I can see it should be an extremely rare case, but it's possible. Also I don't find any other valuable reasons for this lock.

In implementation with multiple streams we still have shared between streams `rdc` and `subgrid` variables, that has to be in consisten state with all stream queues. So this is a reason to use shared lock there.

timoninmaxim replied:
Actually, I see a way to avoid shared lock for multi-stream reducer. We don't need such collection as `rcvd` in multi-stream, as there a stream per node. So, this code can be simplified. I'll try to implement it.

timoninmaxim replied:
I simplified a code, now NodePageStream doesn't have own logic for control nodes or page loading, it inherited from PageStream. So, reducers doesn't use shared lock now.

But all streams shared the same lock, as it guards consistency between CacheQueryFuture state, queues and rcvd. In one moment only single stream requesting a pages (others do have `head` to use), so it doesn't affect performance, and work the same way as UnorderedReducer does.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,290 @@
+ * Reducer of distirbuted query that sort result through all nodes. Note that it's assumed that every node
+ * returns pre-sorted collection of data.
+ */
+public class MergeSortDistributedCacheQueryReducer<R> extends UnsortedDistributedCacheQueryReducer<R> {
```
MergeSorted reducer uses per-node Stream, while Unsorted uses single shared Stream and they have different logic of merging results.
Why do you extend Unsorted reducer instead of Abstract one here?
Some common fields can be safely pulled up into the Abstract reduceer.

Also, I'd move all these classes hierarchy to the package  'reducer' and made PageStream/NodeStream to top-level classes, with the package visibility.

timoninmaxim replied:
Done.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/UnsortedDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onPage(@Nullable UUID nodeId, boolean last) {
```
onPage with `boolean last` and onLastPage() looks confusing.
Looks like an architectural lack.
What is the difference between these methods?

timoninmaxim replied:
Merged them to single method.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Reducer for distributed cache query.
+ */
+public interface DistributedCacheQueryReducer<T> extends CacheQueryReducer<T> {
```
Let's have separate classes/interfaces hierarchy for local and distributed queries.
Because they have different contracts for similar methods. E.g. onLastPage() and onPage(nodeId, true);
Having both methods looks ambiguous.

timoninmaxim replied:
Reworked methods within hierarchy, rename some methods. So, now it looks better. Don't think that separation is good idea, as `reducer(): CacheQueryReducer` used within GridCacheQueryFutureAdapter and shared between all types of queries.

AMashenkov replied:
CacheQueryReducer.onPage requires nodeId as a param. 
Was it made just because it is required for DistributedCacheQueryReducer?
Why do we need DistributedCacheQueryReducer at all?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    public boolean onNodeLeft(UUID nodeId);
+
+    /** Blocks while reducer doesn't get first result item for this query. */
+    public void awaitFirstItem() throws InterruptedException;
```
```suggestion
    /** 
    * Blocks current thread until reducer will be ready to return the very first result item for the query. 
    */
    public void awaitInitialization() throws InterruptedException;
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+     * Callback to handle node left.
+     *
+     * @param nodeId Node ID that left a cluster.
+     * @return {@code true} if specified node runs this query.
```
It is unclear, what result will be returned if the node has gone after the last page was received and how it should be interpreted.
Or what if the node has gone, but we already received sufficient data for finishing the query, 
e.g. query has limits, but the user app consumes data slowly. 

I think Reducer should only check internal invariants and mark node as failed.
And caller should recheck 'if the reducer is alive' right after calling this method.

timoninmaxim replied:
Renamed this method and made docs more clear. Now it just checks whether node is running a query in this moment. `onNodeLeft` now handles with `onError` method.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param nodeId Node ID of remote page.
+     * @param last Whether page is last for specified node.
+     * @return Whether page is last for a query.
```
It looks like the method is called when we got a new page and notify the Reducer.
What the purpose to get a flag to the caller back?
Why reducer can't apply additional logic by itself or call a listener (like in true OOP) instead of return magic value.

What the purpose of the `last` parameter? What does `false` value mean and what is the impact to return value?

timoninmaxim replied:
Now there is single onPage method. The `last` param shows if this page is the last page from node `nodeId`. It's required to make some clean up in reducer, if we know that there wouldn't be pages from this node.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @throws IgniteInterruptedCheckedException If thread is interrupted.
+     */
+    public void loadAll() throws IgniteInterruptedCheckedException;
```
Ok. One can forcibly load all the pages, but where will the data be returned or saved?

timoninmaxim replied:
Renamed this method to `requestFullPages`. So this method about requesting whole data at once from remote node.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+    /** Blocks while reducer doesn't get first result item for this query. */
+    public void awaitFirstItem() throws InterruptedException;
+
+    /** Callback that invokes when this query is cancelled. */
+    public void cancel();
```
```suggestion
    /** 
    * Callback is invoked on the query cancellation. 
    */
    public void onCancel();
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java:
```diff
@@ -0,0 +1,58 @@
+
+    /**
+     * Callback that invokes after reducer get last query result page.
+     * Also invokes for failed queries to let reducer know that there won't be new pages.
```
Why there is no onCancel() or onError() callbacks that will cause Exception throwing in `next` method?

timoninmaxim replied:
Added them.

timoninmaxim commented:
Hi @AMashenkov ! Thanks for comments, I refactored code and make it more clear:
1. Move all reducers to separate package, make PageStream, NodePageStream top level classes;
2. Remove custom logic from NodePageStream (now it just wrapper over PageStream with single node) and leverage on PageStream only. It also helps to simplify logic of MergeSortReducer;
3. Reworked CacheQueryReducer interface: added callback methods (onError, onCancel, onFinish), removed/renamed and fixed doc for unclear methods, moving some methods to top level hierarchy class.

Could you please review PR again? Thanks!

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/PageStream.java:
```diff
@@ -0,0 +1,277 @@
+     * List of nodes that respons with cache query result pages. This collection should be cleaned before sending new
+     * cache query request.
+     */
+    private final Collection<UUID> rcvd = new HashSet<>();
```
This class has a child NodeStream, which is per-node stream.
What semantic 'subgrid', 'queue' and 'rcvd' will have in a child class?

timoninmaxim replied:
The semantic is the same. We still have a `subgrid` even it's a single node, and it also has to track `rcvd` to avoid duplicate messages. I think we should reuse this code instead of implementing separate logic for single node stream, that actually will be pretty the same.

korlov42 replied:
why do you decide to split reducer's logic between actual reducer class and this one?

AMashenkov replied:
Why don't use NodeStream only and merge results at higher level?
Maybe NodeStreams can even share same queue for the results.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * @return Head of stream and then clean it.
+     */
+    public R get() {
+        R ret = head;
+
+        head = null;
+
+        return ret;
+    }
```
This method looks useless.
Al the logic could be done in next() method.

timoninmaxim replied:
You're correct. I wanted to separate logic of filling head() and cleaning(), but actually there is no need to do it.

AMashenkov replied:
`
@Override public boolean hasNext() throws IgniteCheckedException {
        if (head != null)
            return true;
       else {
            if (!super.hasNext())
                return false;
            else
                head = super.next();   
       }
}
`

timoninmaxim replied:
Agree, moving fetch value from next() to hasNext() resolves this. Added.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -324,8 +242,8 @@ public void onPage(@Nullable UUID nodeId, @Nullable Collection<?> data, @Nullabl
             if (err != null)
-                synchronized (this) {
-                    enqueue(Collections.emptyList());
+                synchronized (lock) {
+                    reducer().onError();
```
We still need synchronization there, as it's required to change reducer queue atomically with future onDone. If user hands on latch await, we has to count down it after future is done. Otherwise we can miss exception, we have tests for this logic. Please check this comment:
https://github.com/apache/ignite/blob/6b6c32bf1582d50f349f522c04b11d1603049797/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java#L330

We can control future.onDone() within reducer, but I think it isn't responsibility of reducer.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -338,55 +256,83 @@ public void onPage(@Nullable UUID nodeId, @Nullable Collection<?> data, @Nullabl
 
-                synchronized (this) {
-                    enqueue(data);
+                synchronized (lock) {
```
Yes, it requires for syncing `queue`, `rcvd` and `fut.isDone` to avoid duplicate requests sending for nodes that finish query. Also there is required to use shared lock for onError.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryManager.java:
```diff
@@ -786,40 +720,25 @@ private Object convert(Object obj) {
+            fut.qry.query().validate();
 
-            final Object topic = topic(cctx.nodeId(), req.id());
+            addQueryFuture(reqId, fut);
```
Please check, I've done it, but actually code doesn't look clearer I think.

timoninmaxim commented:
Hi @AMashenkov ! I've answered on all comments (2 changes are accepted, other on discussion), could you please have a look?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java:
```diff
@@ -0,0 +1,68 @@
+    public boolean hasNext() throws IgniteCheckedException;
+
+    /**
+     * Offer query result page for reduce. Note that the data collection may contain extension of type T.
+     * In such cases data item contains additional payload for custom reducer logic.
+     *
+     * @param nodeId Node ID that sent this page.
+     * @param data Page data rows.
+     * @param last Whether this page is last for specified {@code nodeId}.
+     * @return {@code true} if this page is final page for query and no more pages are waited, otherwise {@code false}.
```
```suggestion
     * Callback that invoked on receiving a new page. 
     *
     * @param nodeId Node ID that sent this page.
     * @param data Page data rows.
     * @param last Whether this page is last for specified {@code nodeId}.
     * @return {@code true} if this page is final page for query and no more pages are waited, {@code false} otherwise.
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java:
```diff
@@ -0,0 +1,68 @@
+import org.apache.ignite.IgniteCheckedException;
+
+/**
+ * This class is responsible for reducing results of cache query. Query results are delivered with function
+ * {@link #onPage(UUID, Collection, boolean)}. Note that this reducer deeply interacts with corresponding query future
+ * {@link GridCacheQueryFutureAdapter}, so they used the same lock object. It guards reducer pages operations
+ * and the future status. Custom reduce logic is applied within {@link #next()} and {@link #hasNext()}.
```
```suggestion
 * This class is responsible for reducing results of cache query. 
 *
 * Query results are delivered via callback {@link #onPage(UUID, Collection, boolean)}.
 * @see GridCacheQueryFutureAdapter.
```

AMashenkov replied:
Actually, interface don't imply any coupling with the  query future.
If you think that this worth to be mentioned, then abstract class or concrete implementation would be a better place.

timoninmaxim replied:
Actually, shared lock now is part of PageStream only. So we can remove this mention.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public interface DistributedCacheQueryReducer<T> extends CacheQueryReducer<T> {
+    /**
+     * Checks whether cache query still runs on specified node. If a query finished (send all pages) on this node, then
+     * this method has to return {@code false}.
+     *
+     * @param nodeId Node ID.
+     * @return {@code true} if specified node runs this query.
```
```suggestion
     * Checks whether node with provided {@code nodeId} is a map node for the query.
     * Note: if all the pages were received this node, then the method will return {@code false}.
     *
     * @param nodeId Node ID.
     * @return {@code true} if specified node is a map node for the query, {@code false} otherwise.
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java:
```diff
@@ -69,205 +50,43 @@ protected GridCacheDistributedQueryFuture(GridCacheContext<K, V> ctx, long reqId
 
-    /** {@inheritDoc} */
-    @SuppressWarnings("NonPrivateFieldAccessedInSynchronizedContext")
+    /** Fail if a node runs this query left cluster. */
```
```suggestion
    /** {@inheritDoc} */
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryManager.java:
```diff
@@ -101,12 +104,30 @@
         assert cctx.config().getCacheMode() != LOCAL;
 
+        pageRequester = new CacheQueryPageRequester(cctx) {
+            /** {@inheritDoc} */
```
```suggestion
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryManager.java:
```diff
@@ -523,57 +544,31 @@ else if (!cancelled.contains(res.requestId()))
-        final GridCacheDistributedQueryFuture<K, V, ?> fut =
-            new GridCacheDistributedQueryFuture<>(cctx, reqId, qry, nodes);
+        final GridCacheDistributedQueryFuture fut = fields ?
+            new GridCacheDistributedFieldsQueryFuture(cctx, reqId, qry)
+            : new GridCacheDistributedQueryFuture(cctx, reqId, qry);
```
```suggestion
            new GridCacheDistributedFieldsQueryFuture(cctx, reqId, qry) :
            new GridCacheDistributedQueryFuture(cctx, reqId, qry);
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -146,40 +137,50 @@ public GridCacheQueryBean query() {
         }
     }
 
+    /** @return Cache query results reducer. */
```
```suggestion
    /** 
    * @return Cache query results reducer. 
    */
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -146,40 +137,50 @@ public GridCacheQueryBean query() {
      * @throws IgniteCheckedException If query execution failed with an error.
      */
-    public abstract void awaitFirstPage() throws IgniteCheckedException;
+    public abstract void awaitFirstItem() throws IgniteCheckedException;
```
Maybe awaitFirstItemAvailable() will be better.
Actually, we wait here not for the first item/page from remote, but the first item available after the reduction applied.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -306,10 +221,12 @@ else if (capacity > 0) {
      * @param data Page data.
      * @param err Error (if was).
-     * @param finished Finished or not.
+     * @param finished Whether it is the last page for sender node.
```
Maybe 'lastPage' flag would be a better name?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,84 @@
+        this.nodeId = nodeId;
+    }
+
+    /**
+     * @return Head of stream, that is last item returned with {@link #next()}.
+     */
+    public R head() {
```
```suggestion
    /**
    * Peek the stream head for the next item.
    *
    * Note: requires {@link #hasNext()} to be called at first.
    * @return The item will be returned with {@link #next()}.
    */
    R head() {
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Stream over single node.
+ */
+public class NodePageStream<R> extends PageStream<R> {
```
Let's make the class package visible if possible.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,84 @@
+        return head;
+    }
+
+    /**
+     * @return {@code true} If this stream has next row, {@code false} otherwise.
+     */
```
```suggestion
    /** {@inheritDoc} */
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    /**
+     * @return Next item from this stream.
+     */
```
```suggestion
    /** {@inheritDoc} */
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -423,48 +372,35 @@ private R unmaskNull(R obj) {
     }
 
-    /** {@inheritDoc} */
+    /** Forces to stop query, don't care about result.  */
```
Why waiting on future causes query stop?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,40 @@
+     * @param nodeId Node ID.
+     * @return {@code true} if node with provided {@code nodeId} is a map node for the query, {@code false} otherwise.
+     */
+    public boolean mapNode(UUID nodeId);
```
```suggestion
    public boolean isMapNode(UUID nodeId);
```

timoninmaxim replied:
According to code style [1] we should not use "is" as prefix for internal API.

[1] https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-Gettersandsetters

AMashenkov replied:
This is bad codestyle =)
The one have to read javadoc to understand what the method doing. 
It is unclear whether the method checks "the node is map-node" or method "make the node a map-node".
Maybe name "mappedToNode" be better? WDYT?

timoninmaxim replied:
WDYT if we violate this rule and name it `isMapNode`?

timoninmaxim replied:
Renamed to `remoteQueryNode`

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Callback is invoked on the query cancellation.
+     */
+    public void onCancel();
```
It is not actually callback.
```suggestion
    public void cancel();
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Callback that invokes after a query future is done.
+     */
+    public void onFinish();
```
Why do we need this method if we already have onError and onPage with 'last' parameter?

timoninmaxim replied:
There is a comment about it, see invokation [1]. We should release latch after onDone() invokation. Also I've described this in previous conversation: https://github.com/apache/ignite/pull/9081#discussion_r649824494

[1] [GridCacheQueryFutureAdapter:332](https://github.com/apache/ignite/blob/0add93b488ebe7265b2a0f951fcdfd55adbcc01f/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java#L332)

AMashenkov replied:
Is it possible to release the latch in onError() and onPage() method?
This will reduce `future` and `reducer` coupling.

timoninmaxim replied:
It's not possible, as it's required to release latch after `onDone()`. Actually there is a tough coupling of future and reducer. Don't think this is a problem

timoninmaxim replied:
We can move `onFinish()` from `onDone()` and explicity invoke it, but there are about 5-6 places to invoke this method - after every `onDone()`. Looks wrong to use this method such way

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java:
```diff
@@ -69,205 +53,69 @@ protected GridCacheDistributedQueryFuture(GridCacheContext<K, V> ctx, long reqId
+     * Then {@link #get()} can lock on unpredictably long period of time. So we should avoid call it.
+     */
+    private Collection<R> get0() throws IgniteCheckedException {
+        if (!isDone())
+            throw new IgniteIllegalStateException("Unexpected lock on iterator over distributed cache query result.");
```
Why do you allow to get result via get() method, but forbid to wait?
If noone must call get() then it should throws exception in anyway.
If the method is never called and you don't care then overriding is useless.

timoninmaxim replied:
The only place we use get() it's in this class, some lines above.

https://github.com/apache/ignite/blob/0add93b488ebe7265b2a0f951fcdfd55adbcc01f/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java#L82

But I agree, that it a little bit misleading. So I will replace this line of code with `super.get()` and then we can forbid `get()` at all.

korlov42 commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,183 @@
+    private final Map<UUID, NodePageStream<R>> streamsMap;
+
+    /** Array of streams. Used for iteration over result pages. */
+    private final NodePageStream<R>[] streams;
```
array should be replaced with PriorityQueue

timoninmaxim replied:
Now use PriorityQueue instead.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,183 @@
+     * @param off Offset.
+     * @param cmp Comparator.
+     */
+    private static <Z> void bubbleUp(Z[] arr, int off, Comparator<Z> cmp) {
```
Why don't use logN algorithm here?

timoninmaxim replied:
Now use PriorityQueue instead.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryManager.java:
```diff
@@ -101,12 +104,30 @@
 
+        pageRequester = new CacheQueryPageRequester(cctx) {
+            /** {@inheritDoc} */
+            @Override protected void sendLocal(GridCacheQueryRequest req) {
```
This looks ambiguous.
Why don't have just a factory for the requests instead of Requester?

timoninmaxim replied:
I've tried to refactor it, but actually failed to make it better. 

There are 3 events to send a request to other nodes: initial load, ask for next page after finish previous, cancel query. Initiators of those events are independent on each other and the only common place is `reducer`, that is triggered and  performs requests. 

PageRequester's `sendRequest()` also invokes `sendLocal()` that depends much on variables defined in `GridCacheDistributedQueryManager`. The initiators doesn't have dependency on this manager. Then it's required to accumulate all logic of sending request and local handling in single object (`pageRequester`).

Alternatives is make initiators depend on `GridCacheDistributedQueryManager`, that makes code more spaghetti. Split requester on request factory and request sender also doesn't make code clearer. 

I think current solution is OK, it clear enough. The only bad place is `sendLocal` that depends on the manager variables. I think we should not refactor this manager now, as it is out of scope of this ticket. WDYT?

AMashenkov replied:
You can pass a Consumer<Request> instead of creating anonymous class for now or 
even revert back message processing logic and create a ticket for improvement.

timoninmaxim replied:
made sendLocal as Consumer param.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryManager.java:
```diff
@@ -777,133 +724,32 @@ private Object convert(Object obj) {
-                catch (IgniteCheckedException e) {
-                    if (cctx.io().checkNodeLeft(node.id(), e, true)) {
-                        fut.onNodeLeft(node.id());
+                if (c1.payload() == null)
```
AFAIU, only Text queries have a 'payload'. Why is it named 'payload' and not score?
Does 'null' score make sense for the result?

AMashenkov replied:
Comparator instance can be static.

timoninmaxim replied:
I'm going to use this reducer in IndexQuery later. It compares rows by column values, not by score.

timoninmaxim replied:
Make comparator static.

timoninmaxim commented:
@korlov42 @AMashenkov  thanks for review. I pushed an update and answer on some your comments. Could you please have a look again. The re-request review button works only for Andrey, and not for Konstantin. Don't know why :(

timoninmaxim commented:
@AMashenkov, @korlov42  Hi! this is a gentle reminder. Could you please have a look on my changes?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,236 @@
+
+        ClusterNode locNode = null;
+
+        Collection<ClusterNode> rmtNodes = null;
```
```suggestion
```

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,236 @@
+            if (n.id().equals(locNodeId))
+                locNode = n;
+            else {
+                if (rmtNodes == null)
+                    rmtNodes = new ArrayList<>(nodes.size());
+
+                rmtNodes.add(n);
+            }
+        }
+
+        // Request should be sent to remote nodes before the query is processed on the local node.
+        // For example, a remote reducer has a state, we should not serialize and then send
+        // the reducer changed by the local node.
+        if (!F.isEmpty(rmtNodes)) {
+            for (ClusterNode node : rmtNodes) {
```
```suggestion            
```

timoninmaxim replied:
I applied this suggestions, also I unified a code of sending cancel a new pages requests.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/AbstractDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,165 @@
+ * Abstract class for distributed reducer implementations.
+ */
+abstract class AbstractDistributedCacheQueryReducer<R> implements DistributedCacheQueryReducer<R> {
+    /** Collection of streams that don't deliver all pages yet. */
+    private final Map<UUID, NodePageStream<R>> remoteStreams;
```
Why we need a duplicate collection?
Can `streams` be used instead?

timoninmaxim replied:
The `streams` collection's goals are:
1. `hasNext()` and `next()` in a user thread use thread local variable `streams`. 
2. When a stream is fully done (remote pages and local iterator), then we remove it from the `stream` collection.

The `remoteStreams` collection's goals are different:
1. Ignite's utility threads use it for handling new remote pages.
2. When remote pages finish (but not local iterator), then we remove this stream from the collection. 
3. Some background actions (cancel, onNodeLeft) bases on this collection. Then we need track set of remote node ids.

Then the idea was to have `streams` thread locally, and `remoteStreams` concurrently. It simplified a sync logic between collections as they should be independent from each other, also it helps to avoid one more lock in user thread.

It's possible to share single collection for all threads. But we will still have another collection for tracking remote nodes with alive pages. This tracking collection can be a Set<UUID>, and then we will get NodePageStream from the main collection (`stream`). We won't win any from this. I think it's fine to track NodePageStream in 2 independent collectios. WDYT?

BUT. There was a little bug, `onError` uses `streams` and runs in non-user thread, I'll replace it with the `remoteStreams`, and this will be fine.

timoninmaxim replied:
Fix issue with `onError`, also I make `rcvdFirstPage` ordinary HashSet, as we already have a guard on this var.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheEntryWithPayload.java:
```diff
@@ -0,0 +1,68 @@
+import org.apache.ignite.lang.IgniteBiTuple;
+
+/** Represents cache key-value pair and additional payload to compare cache entity by custom rule. */
+public class CacheEntryWithPayload<K, V, P> extends IgniteBiTuple<K, V> {
```
Would you mind to rename this to ScoredCacheEntry or WeightedCacheEntry?

timoninmaxim replied:
Let's name it ScoredCacheEntry, and replace `payload` with `score`. I will expand this class for other cases in future, it will be needed.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/UnsortedDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,70 @@
+            if (page != null && page.hasNext())
+                return true;
+
+            this.streams.remove(s.nodeId());
```
`streams` is not concurrent structore.
However it can be cleared concurrently.

timoninmaxim replied:
`streams` cleared within `cancel()` action. Actually I add this to help GC a little. But I will revert this, as cancel can be also invoked after reaching limit, and in this moment `streams` can be not finished. In current version it doesn't matter, but will be for future optimizations of algorithm.

timoninmaxim replied:
Remove clearing `streams` in cancel().

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,107 @@
+    ) {
+        super(fut, reqId, pageRequester, queueLock, nodes);
+
+        pageCmp = (o1, o2) -> rowCmp.compare(o1.head(), o2.head());
```
Seems, pageCmp can be local variable.

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/MergeSortDistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,107 @@
+    @Override public void cancel() {
+        super.cancel();
+
+        nodePages.clear();
```
Not in sync with the fillWithPage() method.

AMashenkov replied:
Seems, it is possible to lost last pages silently on concurrent cancellation.

timoninmaxim replied:
Hi, this is OK. If a query is cancelled it means that cache query future is already done and we don't need more pages. Cancellation is invoked whether query is finished with timeout, if user closes cursor, if iterator reaches user specified limit, or with any exception that stops this future. So no need to wait those pages.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,236 @@
+            cctx.io().send(nodeId, req, GridIoPolicy.QUERY_POOL);
+        }
+        catch (IgniteCheckedException e) {
+            if (cctx.io().checkNodeLeft(nodeId, e, false)) {
```
I'd like to remain this check. There is a logic of node left. And this code interfere to this situation. I think this behavior should be investigated with some tests. I will create a separate ticket for that.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java:
```diff
@@ -0,0 +1,236 @@
+/**
+ * This class is responsible for sending request for query result pages to remote nodes.
+ */
+public class CacheQueryPageRequester {
```
Done.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/DistributedCacheQueryReducer.java:
```diff
@@ -0,0 +1,71 @@
+    protected final Map<UUID, NodePageStream<T>> pageStreams;
+
+    /** Timestamp of query timeout. */
+    protected final long endTime;
```
Let's move the endTime-related logic out from the distributed reducer to the `NodeStreams` class. When the `endTime` is reached the `NodeStreams#nextPage` should throw an `IgniteTimeoutException`.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,144 @@
+    private NodePage<R> head;
+
+    /** Promise to notify the stream consumer about delivering new page. */
+    private CompletableFuture<UUID> pageReady = new CompletableFuture<>();
```
We should keep not the nodeId promise, but `CompletableFuture<Collection<R>>`.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,144 @@
+    private CompletableFuture<UUID> pageReady = new CompletableFuture<>();
+
+    /** This lock syncs {@link #head} and {@link #noRemotePages}. */
+    private final Object pagesLock = new Object();
```
Let's do synchronization on `this`.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return Query result page.
+     */
+    public NodePage<R> nextPage() {
```
I suggest keeping CompletableFuture<NodePage<R>> the method return type here.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,144 @@
+     */
+    public void cancel() {
+        synchronized (pagesLock) {
+            head = new NodePage<>(nodeId, Collections.emptyList());
```
I suggest completing the future with the `IgniteCancelledException` in case of any errors and change the `cancel` to the `complete(IgniteException e)` or `error(IgniteException e)`.

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java:
```diff
@@ -0,0 +1,144 @@
+    private boolean noRemotePages;
+
+    /** Last delivered page from the stream. */
+    private NodePage<R> head;
```
Do we need it in case of using the ` CompletableFuture<Collection<R>>`?

Mmuzaf commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java:
```diff
@@ -143,43 +126,62 @@ public GridCacheQueryBean query() {
+                }
+            }
+
+            checkError();
```
Should we handle any kind of exceptions here and close the `NodeStream` handlers by calling the `onPageError` method here?
